api/server/runner_httptrigger.go

```diff
@@ -83,56 +82,99 @@ func (trw *triggerResponseWriter) WriteHeader(statusCode int) {
 	}
 	trw.committed = true
 
-	for k, vs := range trw.Header() {
-		if strings.HasPrefix(k, "Fn-Http-H-") {
-			// TODO strip out content-length and stuff here.
-			realHeader := strings.TrimPrefix(k, "Fn-Http-H-")
-			if realHeader != "" { // case where header is exactly the prefix
-				for _, v := range vs {
-					trw.Header().Del(k)
-					trw.Header().Add(realHeader, v)
+	var fnStatus int
+	realHeaders := trw.Header()
+	gwHeaders := make(http.Header, len(realHeaders))
+	for k, vs := range realHeaders {
+		switch {
+		case strings.HasPrefix(k, "Fn-Http-H-"):
+			gwHeader := strings.TrimPrefix(k, "Fn-Http-H-")
+			if gwHeader != "" { // case where header is exactly the prefix
+				gwHeaders[gwHeader] = vs
+			}
+		case k == "Fn-Http-Status":
+			if len(vs) > 0 {
+				statusInt, err := strconv.Atoi(vs[0])
+				if err == nil {
+					fnStatus = statusInt
 				}
 			}
+		case k == "Content-Type", k == "Fn-Call-Id":
+			gwHeaders[k] = vs
 		}
 	}
 
-	gatewayStatus := 200
+	// XXX(reed): this is O(3n)... yes sorry for making it work without making it perfect first
+	for k := range realHeaders {
+		realHeaders.Del(k)
+	}
+	for k, vs := range gwHeaders {
+		realHeaders[k] = vs
+	}
 
+	// XXX(reed): simplify / add tests for these behaviors...
+	gatewayStatus := 200
 	if statusCode >= 400 {
 		gatewayStatus = 502
 	}
+	if fnStatus > 0 {
+		gatewayStatus = fnStatus
+	}
```

Comment:
Is there a gap here whereby a fn could set the fn-http-status code to 200, but then a failure occur resulting in a 500 response, which this block will then set back to 200?

To some degree that is a user problem, "don't set fn-http-status until your sure", but i'm thinking of ways fdk's could fail that might be outside their control that might lead to this scenario.

Reply:
Fixed - made this else if

Reply:
i don't think we really pegged down this behavior well. honestly if the user manages to get an http status out of the function I see no reason to overwrite it with a 5xx for any reason from fn (this was the case without else if)